### PR TITLE
Refactored mwdb-reporter to better comply with 2.6.0 API features

### LINKED ISSUE #7: Blob support
Currently karton-mwdb-reporter doesn't support `type: blob` tasks, which might be useful in some cases.

### PR DESCRIPTION
**Pull request by psrok1**
- Support for blobs
- More consistent metadata adding and logging
- Using new API when MWDB Core >= 2.6.0

Closes #7

### CODE REVIEW

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 1]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
             mwdb.login(mwdb_config["username"], mwdb_config["password"])
         return mwdb
 
-    def _tag_children_blobs(self, config: MWDBConfig) -> None:
-        for item in config.config.values():
-            if type(item) is MWDBBlob:
-                tag = config.family
+    @property
+    def mwdb(self) -> MWDB:
+        if not hasattr(self, "_mwdb"):
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:97]: Is there a specific upside to this over just initializing it in the constructor?

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 2]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
+            setattr(self, "_mwdb", self._get_mwdb())
+        return getattr(self, "_mwdb")
+
+    def _add_tags(self, mwdb_object: MWDBObject, tags: List[str]):
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:101]:
```suggestion
    def _add_tags(self, mwdb_object: MWDBObject, tags: List[str]) -> None:
```

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 3]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
+                    "[%s %s] Already tagged as %s",
+                    mwdb_object.TYPE,
+                    mwdb_object.id,
                     tag,
                 )
-                item.add_tag(tag)
+
+    def _add_attributes(
+        self, mwdb_object: MWDBObject, attributes: Dict[str, List[Any]]
+    ):
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:119]:
```suggestion
    ) -> None:
```

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 4]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
+                self.log.info(
+                    "[%s %s] Adding attribute %s: %s",
+                    mwdb_object.TYPE,
+                    mwdb_object.id,
+                    key,
+                    value,
+                )
+                mwdb_object.add_attribute(key, value)
+
+    def _add_comments(self, mwdb_object: MWDBObject, comments: List[str]):
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:132]:
```suggestion
    def _add_comments(self, mwdb_object: MWDBObject, comments: List[str]) -> None:
```

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 5]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
+        for tag in tags:
+            if tag not in mwdb_object.tags:
                 self.log.info(
-                    "[blob %s] adding family tag '%s' inherited from parent sample",
-                    item.id,
+                    "[%s %s] Adding tag %s", mwdb_object.TYPE, mwdb_object.id, tag
+                )
+                mwdb_object.add_tag(tag)
+            else:
+                self.log.info(
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:110]: We should delete this log or add a similar one for attributes and comments

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 6]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
+            if comment not in mwdb_object.comments:
+                self.log.info(
+                    "[%s %s] Adding comment: %s",
+                    mwdb_object.TYPE,
+                    mwdb_object.id,
+                    repr(comment),
+                )
+                mwdb_object.add_comment(comment)
+
+    def _add_parent(self, mwdb_object: MWDBObject, parent: Optional[MWDBObject]):
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:144]:
```suggestion
    def _add_parent(self, mwdb_object: MWDBObject, parent: Optional[MWDBObject]) -> None:
```

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 7]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
+        attributes: Optional[Dict[str, List[Any]]],
+        comments: Optional[List[str]],
+        karton_id: str,
+    ) -> Tuple[bool, MWDBObject]:
+        """
+        Generic object uploader that submits additional metadata along with the object.
+        Check MWDB Core version to submit them using the most efficient way that depends
+        on currently supported API.
+
+        :param object_getter: Object getter that returns object or None if doesn't exist
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:178]: A little bit :mage:

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 8]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
+                if parent:
+                    metadata.append("parent: " + parent.id)
+                if tags:
+                    metadata.append("tags: " + ",".join(tags))
+                if attributes:
+                    metadata.append("attributes: " + ",".join(attributes.keys()))
+                if not metadata:
+                    metadata_info = "no metadata"
+                else:
+                    metadata_info = "including " + "; ".join(metadata)
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:218]: :thinking:

Comment (nazywam) [reply karton/mwdb_reporter/mwdb_reporter.py:218]: Oh ok, it's just for logging

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 9]
```diff
@@ -71,268 +92,408 @@ def mwdb(self) -> MWDB:
...
-            parent = None
-
-        config = self._upload_config(task, mwdb, family, config_data, parent=parent)
-        return config
 
-    def process_sample(self, task: Task, mwdb: MWDB) -> Optional[MWDBFile]:
-        """
-        Processing of Sample task
+        # Upload original sample
+        sample: Optional[MWDBObject] = None
```
Comment (nazywam) [inline karton/mwdb_reporter/mwdb_reporter.py:436]: Isn't this technically `MWDBFile`?

File: karton/mwdb_reporter/mwdb_reporter.py (hunk 5 above)
Comment (psrok1) [reply karton/mwdb_reporter/mwdb_reporter.py:110]: There is similar one for attributes. Comments are always added incrementally.

File: karton/mwdb_reporter/mwdb_reporter.py (hunk 1 above)
Comment (psrok1) [reply karton/mwdb_reporter/mwdb_reporter.py:97]: :thinking: I was a bit afraid of messing with that constructor but actually it sounds better than lazy initatization

File: karton/mwdb_reporter/mwdb_reporter.py [hunk 10]
```diff
@@ -24,8 +24,9 @@ class MWDBReporter(Karton):
         "payload": {
             "sample": Resource with sample contents
             "parent": optional, Resource with parent sample contents
+                      or identifier (hash) of parent object
```
[the remark this answers is not in the document]
Comment (psrok1) [reply karton/mwdb_reporter/mwdb_reporter.py:27]: Good point, we can choose object by hash also for configurations

[the remark this answers is not in the document]
Comment (psrok1) [reply karton/mwdb_reporter/mwdb_reporter.py:27]: :+1: Done

File: karton/mwdb_reporter/mwdb_reporter.py (hunk 1 above)
Comment (psrok1) [reply karton/mwdb_reporter/mwdb_reporter.py]: Done :+1:

**Pull request merged**